src/features/contracts/contracts-directory.ts

```diff
@@ -8,13 +8,82 @@ import * as stream from 'stream';
 import * as path from 'path';
 import * as os from 'os';
 import type { RepositoryInfo, Contract } from './index';
+import { isValidUrl } from '../../lib/utils';
 
 const pipeline = util.promisify(stream.pipeline);
 const exists = util.promisify(fs.exists);
 const mkdir = util.promisify(fs.mkdir);
 
 const CONTRACTS_BASE_DIR = path.join(os.tmpdir(), 'contracts');
 
+const getDataUri = async (
+	filePath: string,
+	mimeType: 'image/png' | 'image/svg+xml',
+) => {
+	const base64Content = (await fs.promises.readFile(filePath)).toString(
+		'base64',
+	);
+	return `data:${mimeType};base64,${base64Content}`;
+};
+
+// All assets that are stored together with the contract are encoded and stored in a dataurl format.
+const handleLocalAssetUrl = async (assetUrl: string): Promise<string> => {
+	switch (path.extname(assetUrl)) {
+		case '.svg': {
+			return getDataUri(assetUrl, 'image/svg+xml');
+		}
+		case '.png': {
+			return getDataUri(assetUrl, 'image/png');
+		}
+		default:
+			return assetUrl;
+	}
+};
+
+const normalizeAssets = async (
+	contractFilepath: string,
+	assets: Contract['assets'],
+) => {
+	if (!assets || _.isEmpty(assets)) {
+		return assets;
+	}
+
+	const normalizedAssets: Contract['assets'] = {};
+
+	await Promise.all(
+		Object.entries(assets).map(async ([key, asset]) => {
+			if (isValidUrl(asset.url)) {
+				normalizedAssets[key] = asset;
+				return;
+			}
+
+			// Convert from relative to absolute path for the asset file and make sure it doesn't try to access files outside of the contract folder.
+			const contractDir = path.dirname(contractFilepath);
+			const normalizedUrl = path.resolve(contractDir, asset.url);
```

Comment:
You need to check that `asset.url` does not include anything like a `..` which might escape the contracts directory and also that no symlinks to outside of the directory can be followed

Reply:
@Page- doesn't the check below this line suffice? This will resolve it into an absolute URL, and then we check if that absolute URL is within the contracts repo.

Reply:
Ah, yeah it should cover the `..` case, not sure about symlinks though

Reply:
I don't think there is a way for this to be a symlink since it's just a string path, not an actual file that symlinks to something else.

Reply:
You can have directory symlinks @sradevski and the path is read from the filesystem so there is definitely the possibility to resolve those symlinks, the final .svg or .png could also be a symlink itself couldn't it? Overall I'd prefer to be extra careful with anything that can read from the filesystem because it'd be really bad if someone managed to grab config variables

Reply:
Fair point, even though we control the contract repos it's better to err on the safe side, I agree. Added a check for symlinks as well